File: app/src/main/java/com/liweiyap/narradir/ui/NarradirFragmentBase.java
package com.liweiyap.narradir.ui;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavController;
import androidx.navigation.fragment.NavHostFragment;

import com.liweiyap.narradir.util.INarradirControl;
import com.liweiyap.narradir.util.NarradirViewModel;

public abstract class NarradirFragmentBase extends Fragment
{
    @Override
    public void onAttach(@NonNull Context context)
    {
        super.onAttach(context);

        if (!(context instanceof INarradirControl))
        {
            throw new RuntimeException(
                "NarradirFragmentBase::onAttach(): " +
                    "Programming error: Context is not NarradirControl."
            );
        }

        mNarradirControl = (INarradirControl) context;
    }

    @Override
    public void onDetach()
    {
        super.onDetach();
        mNarradirControl = null;
    }

    protected INarradirControl getNarradirControl()
    {
        return mNarradirControl;
    }

    @Nullable
    protected NarradirViewModel getViewModel()
    {
        if (mNarradirControl == null)
        {
            return null;
        }

        return mNarradirControl.getViewModel();
    }

    protected void addSoundToPlayOnButtonClick(final IObserverListener btn)
    {
        if (btn == null)
        {
            return;
        }

        btn.addOnClickObserver(() -> {
            INarradirControl narradirControl = getNarradirControl();
            if (narradirControl != null)
            {
                narradirControl.playClickSound();
            }
        });
    }

    protected void navigateUp(final int steps)
    {
        NavController navController = NavHostFragment.findNavController(this);
        for (int step = 0; step < steps; ++step)
        {
            navController.navigateUp();
        }
    }

    private INarradirControl mNarradirControl;
}

File: app/src/main/java/com/liweiyap/narradir/ui/NarradirFragmentBase.kt
package com.liweiyap.narradir.ui

import android.content.Context
import android.os.Bundle
import androidx.annotation.IdRes

import androidx.fragment.app.Fragment

import com.liweiyap.narradir.util.INarradirControl
import com.liweiyap.narradir.util.NarradirViewModel
import com.liweiyap.narradir.util.SafeNavigator

abstract class NarradirFragmentBase: Fragment() {
    override fun onAttach(context: Context) {
        super.onAttach(context)

        if (context !is INarradirControl) {
            throw RuntimeException(
                "NarradirFragmentBase::onAttach(): " +
                    "Programming error: Context is not INarradirControl.")
        }

        mNarradirControl = context
    }

    override fun onDetach() {
        super.onDetach()
        mNarradirControl = null
    }

    protected open fun addSoundToPlayOnButtonClick(btn: IObserverListener?) {
        btn?.addOnClickObserver {
            mNarradirControl?.playClickSound()
        }
    }

    protected fun navigateToFragment(@IdRes resId: Int) {
        SafeNavigator.navigate(fragment = this, resId = resId)
    }

    protected fun navigateToFragment(@IdRes resId: Int, args: Bundle?) {
        SafeNavigator.navigate(fragment = this, resId = resId, args = args)
    }

    protected fun navigateUp(steps: Int) {
        SafeNavigator.navigateUp(fragment = this, steps = steps)
    }

    protected var mNarradirControl: INarradirControl? = null
        private set

    protected val viewModel: NarradirViewModel?
        get() = mNarradirControl?.getViewModel()
}